File: apps/ios/PhrenKit/Sources/PhrenKit/Markdown/ReviewFile.swift
import Foundation

/// Parser + mutator for a project's review.md (the review queue).
///
/// Transcribes `readReviewQueue` / `parseQueueLine` / `approveQueueItem` /
/// `rejectQueueItem` / `editQueueItem` (packages/cli/src/data/access.ts:609-749)
/// and the queue text normalization from governance/policy.ts:706-741.
public struct ReviewFile: Sendable {
    public private(set) var content: String

    public init(content: String) {
        self.content = content
    }

    // MARK: - Queue text normalization (governance/policy.ts)

    /// policy.ts:19
    static let maxQueueEntryLength = 500

    /// policy.ts:710 `cleanQueueEntryText`
    static func cleanQueueEntryText(_ raw: String) -> String {
        var s = JSRegex(#"\r\n?"#).replaceAll(raw, with: "\n")
        s = s.replacingOccurrences(of: "\0", with: " ")
        // policy.ts uses /<!--[\s\S]*?-->/g (dot-matches-newline variant).
        s = JSRegex(#"<!--[\s\S]*?-->"#).replaceAll(s, with: " ")
        s = JSRegex(#"\\[nrt]"#).replaceAll(s, with: " ")
        s = s.replacingOccurrences(of: "\\\"", with: "\"")
        s = s.replacingOccurrences(of: "\\\\", with: "\\")
        s = JSRegex(#"\n+"#).replaceAll(s, with: " ")
        s = s.collapsedWhitespace
        return s.jsTrimmed
    }

    /// policy.ts:723 `normalizeQueueEntryText` with `{truncate: true}`.
    /// Lengths are measured in UTF-16 units to match JS `.length`/`.slice` —
    /// otherwise emoji shift the truncation boundary and break byte-identity.
    static func normalizeQueueEntryText(_ raw: String) -> String {
        let cleaned = cleanQueueEntryText(raw)
        if cleaned.utf16.count <= maxQueueEntryLength { return cleaned }
        let units = Array(cleaned.utf16.prefix(maxQueueEntryLength - 1))
        let sliced = String(utf16CodeUnits: units, count: units.count)
        return trimEnd(sliced) + "…"
    }

    // MARK: - Parse (access.ts:609 parseQueueLine, 631 readReviewQueue)

    struct ParsedQueueLine {
        var date: String?
        var text: String
        var confidence: Double?
        var machine: String?
        var model: String?
    }

    static func parseQueueLine(_ line: String) -> ParsedQueueLine {
        let dated = JSRegex(#"^- \[(\d{4}-\d{2}-\d{2})\]\s*(.+)$"#)
        var date: String?
        var rawText: String
        if let m = dated.firstMatch(in: line),
           let d = JSRegex.substring(line, m, 1),
           let t = JSRegex.substring(line, m, 2) {
            date = d
            rawText = t
        } else {
            rawText = JSRegex(#"^-\s+"#).replaceFirst(line, with: "").jsTrimmed
        }
        let confidence = JSRegex(#"\[confidence\s+([01](?:\.\d+)?)\]"#, caseInsensitive: true)
            .group(rawText).flatMap(Double.init)
        let source = parseSourceComment(line)
        let withoutConfidence = JSRegex(#"\s*\[confidence\s+[01](?:\.\d+)?\]"#, caseInsensitive: true)
            .replaceAll(rawText, with: "").jsTrimmed
        let text = normalizeQueueEntryText(withoutConfidence)
        return ParsedQueueLine(
            date: date, text: text, confidence: confidence,
            machine: source?.machine, model: source?.model
        )
    }

    public func parse() -> [QueueItem] {
        guard !content.isEmpty else { return [] }
        var items: [QueueItem] = []
        var section: QueueItem.Section = .review
        var index = 1

        for line in content.components(separatedBy: "\n") {
            let trimmed = line.jsTrimmed
            if let heading = JSRegex(#"^##\s+(.+?)[\s]*$"#, caseInsensitive: true).group(trimmed) {
                let token = heading.collapsedWhitespace.jsTrimmed.lowercased()
                if token == "review" { section = .review; continue }
                if token == "stale" { section = .stale; continue }
                if token == "conflicts" { section = .conflicts; continue }
            }
            guard line.hasPrefix("- ") else { continue }

            let parsed = Self.parseQueueLine(line)
            let risky = section != .review || (parsed.confidence.map { $0 < 0.7 } ?? false)
            items.append(QueueItem(
                id: "M\(index)",
                section: section,
                date: parsed.date ?? "unknown",
                text: parsed.text,
                line: line,
                confidence: parsed.confidence,
                risky: risky,
                machine: parsed.machine,
                model: parsed.model
            ))
            index += 1
        }
        return items
    }

    // MARK: - Line operations (access.ts:673-749)

    /// access.ts:674 `withQueueLineOp` — locate by trimmed-line equality.
    private func lineIndex(of lineText: String, in lines: [String]) throws -> Int {
        guard let idx = lines.firstIndex(where: { $0.jsTrimmed == lineText.jsTrimmed }) else {
            throw PhrenKitError.notFound("Queue item not found.")
        }
        return idx
    }

    /// Remove a queue line, touching nothing else.
    ///
    /// This is only the review.md half of approve or reject. Approve must also
    /// write the finding when it is not already there: `phren extract` queues
    /// every candidate below autoAcceptThreshold without writing it to
    /// FINDINGS.md, so for those the queue line is the only copy. SyncEngine
    /// composes this with `FindingsFile` (see `computeEdits`).
    public mutating func dequeue(lineText: String) throws {
        var lines = content.components(separatedBy: "\n")
        let idx = try lineIndex(of: lineText, in: lines)
        lines.remove(at: idx)
        content = FindingsFile.normalizeWrite(lines)
    }

    /// The review.md half of `approveQueueItemDetailed` (access.ts).
    public mutating func approve(lineText: String) throws {
        try dequeue(lineText: lineText)
    }

    /// The review.md half of `rejectQueueItem` (access.ts:709). The caller
    /// composes this with `FindingsFile.remove` using `findingsTextFor(lineText:)`,
    /// tolerating a not-found finding exactly like the CLI does.
    public mutating func reject(lineText: String) throws {
        try dequeue(lineText: lineText)
    }

    /// Capture provenance recorded on the queue line, if any. A promoted
    /// finding carries where the observation came from, not the device that
    /// approved it, as `approveQueueItemDetailed` does.
    public static func capturedProvenanceFor(lineText: String) -> FindingProvenance? {
        parseSourceComment(lineText)
    }

    /// The date the item was queued, recorded on a promoted finding as
    /// `<!-- phren:queued "YYYY-MM-DD" -->`.
    public static func queuedDateFor(lineText: String) -> String? {
        parseQueueLine(lineText).date
    }

    /// The parsed queue text used as the FINDINGS.md match needle for
    /// reject/edit (`parseQueueLine(lineText).text` in access.ts:717,732).
    public static func findingsTextFor(lineText: String) -> String {
        parseQueueLine(lineText).text
    }

    /// The review.md half of `editQueueItem` (access.ts:728) — rewrites the
    /// line preserving the `- [date] ` prefix. Returns the new-text needle for
    /// the tolerant FINDINGS.md edit.
    @discardableResult
    public mutating func edit(lineText: String, newText: String) throws -> String {
        let trimmed = JSRegex(#"[\r\n]+"#).replaceAll(newText, with: " ").jsTrimmed
        guard !trimmed.isEmpty else { throw PhrenKitError.emptyInput("New text cannot be empty.") }

        var lines = content.components(separatedBy: "\n")
        let idx = try lineIndex(of: lineText, in: lines)
        if let date = JSRegex(#"^- \[(\d{4}-\d{2}-\d{2})\]\s*"#).group(lines[idx]) {
            lines[idx] = "- [\(date)] \(trimmed)"
        } else {
            lines[idx] = "- \(trimmed)"
        }
        content = FindingsFile.normalizeWrite(lines)
        return trimmed
    }
}

File: apps/android/phrenkit/src/main/kotlin/com/phren/kit/ReviewFile.kt
package com.phren.kit

/**
 * Parser + mutator for a project's review.md (the review queue). Port of
 * ReviewFile.swift, transcribing packages/cli/src/data/access.ts:609-749 and
 * the queue text normalization from governance/policy.ts:706-741.
 */
class ReviewFile(content: String) {
    var content: String = content
        private set

    data class ParsedQueueLine(
        val date: String?,
        val text: String,
        val confidence: Double?,
        val machine: String?,
        val model: String?,
    )

    fun parse(): List<QueueItem> {
        if (content.isEmpty()) return emptyList()
        val items = mutableListOf<QueueItem>()
        var section = QueueItem.Section.REVIEW
        var index = 1
        for (line in content.split("\n")) {
            val heading = HEADING.group(line.jsTrimmed)
            if (heading != null) {
                when (heading.collapsedWhitespace.jsTrimmed.lowercase()) {
                    "review" -> { section = QueueItem.Section.REVIEW; continue }
                    "stale" -> { section = QueueItem.Section.STALE; continue }
                    "conflicts" -> { section = QueueItem.Section.CONFLICTS; continue }
                }
            }
            if (!line.startsWith("- ")) continue
            val parsed = parseQueueLine(line)
            val risky = section != QueueItem.Section.REVIEW || (parsed.confidence?.let { it < 0.7 } ?: false)
            items += QueueItem(
                id = "M$index",
                section = section,
                date = parsed.date ?: "unknown",
                text = parsed.text,
                line = line,
                confidence = parsed.confidence,
                risky = risky,
                machine = parsed.machine,
                model = parsed.model,
            )
            index++
        }
        return items
    }

    /** access.ts:674 `withQueueLineOp` — locate by trimmed-line equality. */
    private fun lineIndex(lineText: String, lines: List<String>): Int {
        val idx = lines.indexOfFirst { it.jsTrimmed == lineText.jsTrimmed }
        if (idx < 0) throw PhrenKitError.NotFound("Queue item not found.")
        return idx
    }

    /** access.ts:700 `approveQueueItem` — remove the line from review.md only. */
    fun approve(lineText: String) {
        val lines = content.split("\n").toMutableList()
        lines.removeAt(lineIndex(lineText, lines))
        content = FindingsFile.normalizeWrite(lines)
    }

    /**
     * The review.md half of `rejectQueueItem` (access.ts:709). The caller
     * composes this with `FindingsFile.remove` using [findingsTextFor].
     */
    fun reject(lineText: String) = approve(lineText)

    /**
     * The review.md half of `editQueueItem` (access.ts:728) — rewrites the line
     * preserving the `- [date] ` prefix. Returns the new-text needle.
     */
    fun edit(lineText: String, newText: String): String {
        val trimmed = NEWLINES.replaceAll(newText, " ").jsTrimmed
        if (trimmed.isEmpty()) throw PhrenKitError.EmptyInput("New text cannot be empty.")
        val lines = content.split("\n").toMutableList()
        val idx = lineIndex(lineText, lines)
        val date = DATE_PREFIX.group(lines[idx])
        lines[idx] = if (date != null) "- [$date] $trimmed" else "- $trimmed"
        content = FindingsFile.normalizeWrite(lines)
        return trimmed
    }

    companion object {
        /** policy.ts:19 */
        const val MAX_QUEUE_ENTRY_LENGTH = 500

        private val HEADING = JSRegex("""^##\s+(.+?)[\s]*$""", caseInsensitive = true)
        private val CRLF = JSRegex("""\r\n?""")
        private val COMMENT_DOTALL = JSRegex("""<!--[\s\S]*?-->""")
        private val ESCAPED_WS = JSRegex("""\\[nrt]""")
        private val NEWLINE_RUN = JSRegex("""\n+""")
        private val DATED = JSRegex("""^- \[(\d{4}-\d{2}-\d{2})\]\s*(.+)$""")
        private val LEADING_BULLET = JSRegex("""^-\s+""")
        private val CONFIDENCE = JSRegex("""\[confidence\s+([01](?:\.\d+)?)\]""", caseInsensitive = true)
        private val CONFIDENCE_STRIP = JSRegex("""\s*\[confidence\s+[01](?:\.\d+)?\]""", caseInsensitive = true)
        private val NEWLINES = JSRegex("""[\r\n]+""")
        private val DATE_PREFIX = JSRegex("""^- \[(\d{4}-\d{2}-\d{2})\]\s*""")

        /** policy.ts:710 `cleanQueueEntryText` */
        fun cleanQueueEntryText(raw: String): String {
            var s = CRLF.replaceAll(raw, "\n")
            s = s.replace("\u0000", " ")
            s = COMMENT_DOTALL.replaceAll(s, " ")
            s = ESCAPED_WS.replaceAll(s, " ")
            s = s.replace("\\\"", "\"")
            s = s.replace("\\\\", "\\")
            s = NEWLINE_RUN.replaceAll(s, " ")
            s = s.collapsedWhitespace
            return s.jsTrimmed
        }

        /**
         * policy.ts:723 `normalizeQueueEntryText` with `{truncate: true}`.
         * Kotlin strings are UTF-16 like JS, so `.length`/`.substring` match
         * the CLI's truncation boundary exactly (lone surrogates included).
         */
        fun normalizeQueueEntryText(raw: String): String {
            val cleaned = cleanQueueEntryText(raw)
            if (cleaned.length <= MAX_QUEUE_ENTRY_LENGTH) return cleaned
            return trimEnd(cleaned.substring(0, MAX_QUEUE_ENTRY_LENGTH - 1)) + "…"
        }

        /** access.ts:609 `parseQueueLine` */
        fun parseQueueLine(line: String): ParsedQueueLine {
            val m = DATED.firstMatch(line)
            val date: String?
            val rawText: String
            if (m != null) {
                date = m.group(1)
                rawText = m.group(2)
            } else {
                date = null
                rawText = LEADING_BULLET.replaceFirst(line, "").jsTrimmed
            }
            val confidence = CONFIDENCE.group(rawText)?.toDoubleOrNull()
            val source = parseSourceComment(line)
            val withoutConfidence = CONFIDENCE_STRIP.replaceAll(rawText, "").jsTrimmed
            return ParsedQueueLine(date, normalizeQueueEntryText(withoutConfidence), confidence, source?.machine, source?.model)
        }

        /** The parsed queue text used as the FINDINGS.md match needle (access.ts:717,732). */
        fun findingsTextFor(lineText: String): String = parseQueueLine(lineText).text
    }
}
